src/meta/KVBasedGflagsManager.cpp

```diff
@@ -0,0 +1,141 @@
+    LOG(FATAL) << "Unimplement!";
+    return Status::NotSupported();
+}
+
+void KVBasedGflagsManager::getGflagsModule() {
+    // get current process according to gflags pid_file
+    gflags::CommandLineFlagInfo pid;
+    if (gflags::GetCommandLineFlagInfo("pid_file", &pid)) {
+        auto defaultPid = pid.default_value;
+        if (defaultPid.find("nebula-graphd") != std::string::npos) {
+            module_ = cpp2::ConfigModule::GRAPH;
+        } else if (defaultPid.find("nebula-storaged") != std::string::npos) {
+            module_ = cpp2::ConfigModule::STORAGE;
+        } else if (defaultPid.find("nebula-metad") != std::string::npos) {
+            module_ = cpp2::ConfigModule::META;
+        } else {
+            LOG(ERROR) << "Should not reach here";
+        }
+    } else {
+        LOG(ERROR) << "Should not reach here";
+    }
+}
+
+Status KVBasedGflagsManager::registerGflags() {
+    // based on kv, so we will retry until register config successfully
+    bool succeeded = false;
+    while (!succeeded) {
+        std::vector<kvstore::KV> data;
+        folly::SharedMutex::WriteHolder wHolder(LockUtils::configLock());
+        for (const auto& item : gflagsDeclared_) {
+            auto module = item.get_module();
+            auto name = item.get_name();
+            auto type = item.get_type();
+            auto mode = item.get_mode();
+            auto value = item.get_value();
+
+            std::string configKey = MetaServiceUtils::configKey(module, name);
+            std::string configValue;
+            // ignore config which has been registered before
+            auto ret = kvstore_->get(kDefaultSpaceId, kDefaultPartId, configKey, &configValue);
+            if (ret == kvstore::ResultCode::SUCCEEDED) {
+                continue;
+            }
+            configValue = MetaServiceUtils::configValue(type, mode, value);
+            data.emplace_back(configKey, configValue);
+        }
+
+        if (!data.empty()) {
+            // still have some gflags unregistered, try to write
+            folly::Baton<true, std::atomic> baton;
+            kvstore_->asyncMultiPut(kDefaultSpaceId, kDefaultPartId, std::move(data),
+                                    [&] (kvstore::ResultCode code) {
+                if (code == kvstore::ResultCode::SUCCEEDED) {
+                    succeeded = true;
+                } else if (code == kvstore::ResultCode::ERR_LEADER_CHANGED) {
+                    auto leader = kvstore_->partLeader(kDefaultSpaceId, kDefaultPartId);
+                    if (ok(leader)) {
+                        auto addr = value(std::move(leader));
+                        // HostAddr(0, 0) means no leader exists
+                        if (addr != HostAddr(0, 0)) {
+                            succeeded = true;
+                        }
+                    }
+                }
+                baton.post();
+            });
+            baton.wait();
```

Comment:
it is dangerous to call wait after current thread hold a write lock.

Reply:
It seems we want make sure that we register succeed, and it it called only once when we init.